frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,330 @@
+  const forceRerender = (): void => {
+    setRerender(prev => prev + 1)
+  }
+  const [progressTime, setProgressTime] = useState(STARTING_TIME_STRING)
+  const [recordingTime, setRecordingTime] = useState(STARTING_TIME_STRING)
+  const [shouldUpdatePlaybackTime, setShouldUpdatePlaybackTime] =
+    useState(false)
+  const [hasNoMicPermissions, setHasNoMicPermissions] = useState(false)
+
+  const widgetId = element.id
+  const widgetFormId = element.formId
+
+  const uploadTheFile = useCallback(
+    (file: File) => {
+      uploadFiles({
+        files: [file],
+        uploadClient,
+        widgetMgr,
+        widgetInfo: { id: widgetId, formId: widgetFormId },
+      }).then(({ successfulUploads }) => {
+        const upload = successfulUploads[0]
+        if (upload && upload.fileUrl.deleteUrl) {
+          setDeleteFileUrl(upload.fileUrl.deleteUrl)
+        }
+      })
+    },
+    [uploadClient, widgetMgr, widgetId, widgetFormId]
+  )
+
+  useEffect(() => {
+    // this first part is to ensure we prompt for getting the user's media devices
+    navigator.mediaDevices
+      .getUserMedia({ audio: true })
+      .then(() => {
+        RecordPlugin.getAvailableAudioDevices().then(devices => {
+          setAvailableAudioDevices(devices)
+          if (devices.length > 0) {
+            setActiveAudioDeviceId(devices[0].deviceId)
+          }
+        })
+      })
+      .catch(_err => {
+        setHasNoMicPermissions(true)
+      })
+  }, [])
+
+  const initializeWaveSurfer = useCallback(() => {
+    if (waveSurferRef.current === null) return
+
+    const ws = WaveSurfer.create({
+      container: waveSurferRef.current,
+      waveColor: theme.colors.primary,
+      progressColor: theme.colors.bodyText,
+      height: WAVEFORM_HEIGHT - 2 * WAVEFORM_PADDING,
+      barWidth: BAR_WIDTH,
+      barGap: BAR_GAP,
+      barRadius: BAR_RADIUS,
+      cursorWidth: CURSOR_WIDTH,
+    })
+
+    ws.on("timeupdate", time => {
+      setProgressTime(formatTime(time * 1000)) // get from seconds to milliseconds
+    })
+
+    ws.on("pause", () => {
+      forceRerender()
+    })
+
+    const rp = ws.registerPlugin(
+      RecordPlugin.create({
+        scrollingWaveform: false,
+        renderRecordedAudio: true,
+      })
+    )
+
+    rp.on("record-end", blob => {
+      const url = URL.createObjectURL(blob)
+      setRecordingUrl(url)
+
+      const file = new File([blob], "audio.wav", { type: blob.type })
+      uploadTheFile(file)
+
+      ws.setOptions({
+        // This color is hardcoded because I've spent an hour trying to get it to work
+        // with the theme and I'm giving up for now. The complications arise due to the
+        // way wavesurfer is using these colors. Specifically, the progress color "tints" the wave color
+        // and the opacity of this wave color is not being respected, so to make things work for now hardcoding
+        // this color that looks "okay" on both themes.
+        waveColor: "#A5A5AA",
```

Comment:
Okay, I did some more looking into this. Here's the problem:
fadedText40 -- the faded part comes from the fact that transparency is used
progressWaveColor -- the wavesurfer property is a _tint_ not a color that just replaces it

the lower opacity on fadedText40 carries over to the tint making it not possible to tint it black how we want. (even if i specify "#000")

so in order to fix this, we need a color that has 100% opacity that comes out to be the same color as fadedText40 does on the background its on. does that make sense?

Reply:
so are you suggesting taking the fadedText40 and using that blend function to blend it with the background and then setting that resulting color as the waveColor?

Reply:
Yep, that sounds like using `waveColor = blend(fadedText40, background)` could work and might be worth a try. Btw. the blend code above is from [glide-data-grid](https://github.com/glideapps/glide-data-grid) (the underlying data grid library of Streamlit) where it is used a lot for similar aspects. But not 100% sure if it works correctly for the case here. If it doesn't work correctly, I can also take another look on this later with the advanced theming project.

Reply:
whew... for a second i thought you just casually pulled that algo out of thin air 😅 . let me try it and see what happens

Reply:
so i got this to work, but only by putting Math.round(...) around the ro, bo, and go lines because it was trying to convert floats in the rgba to hex and struggling. i think this color should be close enough that it's good enough for now? wdyt?

Reply:
Yep, I think this sounds reasonable 👍 Thanks for taking another look!